scanpy/_utils/__init__.py

```diff
@@ -560,6 +569,160 @@ def _elem_mul_dask(x: DaskArray, y: DaskArray) -> DaskArray:
     return da.map_blocks(elem_mul, x, y)
 
 
+Scaling_T = TypeVar("Scaling_T", DaskArray, np.ndarray)
+
+
+def broadcast_axis(divisor: Scaling_T, axis: Literal[0, 1]) -> Scaling_T:
+    divisor = np.ravel(divisor)
+    if axis:
+        return divisor[None, :]
+    return divisor[:, None]
+
+
+@singledispatch
+def axis_scale(
+    X: sparse.spmatrix,
+    scaling_array,
+    axis: Literal[0, 1],
+    *,
+    out: sparse.spmatrix | None = None,
+) -> sparse.spmatrix:
+    if X is out:
+        if axis == 0:
+            sparsefuncs.inplace_row_scale(X, np.ravel(scaling_array))
+            return X
+        sparsefuncs.inplace_column_scale(X, np.ravel(scaling_array))
+        return X
+
+    row_scale = axis == 0
+    column_scale = axis == 1
+    # Same as `inplace_{column,row}_scale`, but not in place
+    if X.format == "csr":
+        if row_scale:
+            new_data = X.data * np.repeat(scaling_array, np.diff(X.indptr))
+        elif column_scale:
+            new_data = X.data * scaling_array.take(X.indices, mode="clip")
+        return sparse.csr_matrix((new_data, X.indices, X.indptr), shape=X.shape)
```

Comment:
```suggestion
        return sparse.csr_matrix((new_data, X.indices.copy(), X.indptr.copy()), shape=X.shape)
```

We should probably be making copies of these if we're not doing the operation inplace.